src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -0,0 +1,620 @@
+    // onExecutionCompleted handle both types of instructions.
+
+    /**
+     * This function serves as the event handler for when the CPU has provided
+     * us with the data we requested to be fetched.
+     *
+     * This function should be called after onPCTranslated().
+     */
+    void onInstDataFetched(std::weak_ptr<InflightInst> inst,
+                              const TheISA::MachInst fetch_data);
+
+    /**
+     * This function serves as the event handler for when the CPU has completed
+     * the translation we requested for fetching purposes.
+     *
+     * This function should be called after onBeginFetch().
+     *
+     * @param fault If the translation request generated a fault, it should be
+     *  returned through this parameter.
+     * @param req A reference to a request object which should hold the
+     *  translated physical address.
+     */
+    void onPCTranslated(std::weak_ptr<InflightInst> inst, Fault fault,
+                        const RequestPtr& req);
+
+    bool outstandingFault()
+    { return earliestOutstandingFaultingInstNum > lastCommittedInstNum; }
+
+    /**
+     * This function detects any dependencies that a new instruction has, if it
+     * were to be issued at the current moment, considering the state of all
+     * currently active InflightInsts. The dependencies should be added to this
+     * instruction via its addDependency() interface function.
+     *
+     * @param inst The instruction for which we want to detect dependencies.
+     */
+    void populateDependencies(std::shared_ptr<InflightInst> inst);
+
+    /**
+     * This function iterates through the registers that this instruction
+     * writes and associates the registers with the corresponding instructions
+     * in order to streamline dependency tracking. Like a scoreboard.
+     *
+     * @param inst The instruction for which we want to track uses.
+     */
+    void populateUses(std::shared_ptr<InflightInst> inst);
+
+    /**
+     * This function checks if the thread should be active at this moment, and
+     * whether or not we have a reasonable PC value ready to be able to
+     * initiate another fetch at this time.
+     *
+     * @return Whether we should start a new fetch at this point in time.
+     */
+    bool hasNextPC();
+
+    // END Internal functions
+
+  public:
+
+    // Fullsystem mode constructor
+    SDCPUThread(SimpleDataflowCPU* cpu_, ThreadID tid_, System* system_,
+                     BaseTLB* itb_, BaseTLB* dtb_, TheISA::ISA* isa_,
+                     bool use_kernel_stats_ = true);
+
+    // Non-fullsystem constructor
+    SDCPUThread(SimpleDataflowCPU* cpu_, ThreadID tid_, System* system_,
+                     Process* process_, BaseTLB* itb_, BaseTLB* dtb_,
+                     TheISA::ISA* isa_);
+
+    // May need to define move constructor, due to how SimpleThread is defined,
+    // if we want to hold instances of these in a vector instead of pointers
+    // gotten from new.
```

Comment:
Ok. I see why you have to have pointers in the vector now. Ignore those comments.